### PR TITLE
Fix filters parse and display

### LINKED ISSUE #5060: Missing filter status?
It appears some Filter statuses are not parsed/rendered correctly? 

See cancer demo case.

Screenshot 2024-11-22 at 09 27 42
Screenshot 2024-11-22 at 09 34 35

**Describe the bug**

**To Reproduce**

**Expected behavior**

**Screenshots**

**Additional context**

### PR DESCRIPTION
**Pull request by dnil**
This PR adds a functionality or fixes a bug.

- Fix #5060 

Bit of a difference already without touching parsing:
Screenshot 2024-11-22 at 10 44 30

Then with parsing fixed:
Screenshot 2024-11-22 at 14 21 35

Screenshot 2024-11-22 at 14 34 12

I also added a filters badge to the variant page for good measure:
Screenshot 2024-11-22 at 14 49 06

<details>
<summary>Testing on cg-vm1 server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. Make sure the PR is pushed and available on [Docker Hub](https://hub.docker.com/repository/docker/clinicalgenomics/scout-server-stage)
1. Fist book your testing time using the Pax software available at [https://pax.scilifelab.se/](https://pax.scilifelab.se). The resource you are going to call dibs on is `scout-stage` and the server is `cg-vm1`.
1. `ssh <USER.NAME>@cg-vm1.scilifelab.se`
1. `sudo -iu hiseq.clinical`
1. `ssh localhost`
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `podman ps`
1. Stop the service with current deployed branch: `systemctl --user stop scout.target`
1. Start the scout service with the branch to test: `systemctl --user start scout@<this_branch>`
1. Make sure the branch is deployed: `systemctl --user status scout.target`
1. After testing is done, repeat procedure at [https://pax.scilifelab.se/](https://pax.scilifelab.se), which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>

<details>
<summary>Testing on hasta server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. `ssh <USER.NAME>@hasta.scilifelab.se`
1. Book your testing time using the Pax software. `us; paxa -u <user> -s hasta -r scout-stage`. You can also use the WSGI Pax app available at [https://pax.scilifelab.se/](https://pax.scilifelab.se).
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `conda activate S_scout; pip freeze | grep scout-browser`
1. Deploy the branch to test: `bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_scout -t scout -b <this_branch>`
1. Make sure the branch is deployed: `us; scout --version`
1. After testing is done, repeat the `paxa` procedure, which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>

**How to test**:
1. check a cancer case and an RD case, note filter tags missing and tags saying Pass when they should say Filtered. 
2. apply patch
3. note improvements with old cases, e.g. some new danger badges on variantS, and the filters badge on the variant page. And in particular no new issues with them.
4. for full effect, load/update the cases with the patch and note good tags both for callers and filters

**Expected outcome**:
The functionality should be working
Take a screenshot and attach or copy/paste the output.

**Review:**
- [ ] code approved by
- [ ] tests executed by

### CODE REVIEW

**Comment by dnil** [comment]
Ok, with that out of the world, I would say we are not formally bugged at least not based on the demo case situation.
Screenshot 2024-11-22 at 10 53 13
Since the `set` tag is used without a "filteredIn", it is hard to fault the parsing here. This looks like more like a merging/pipeline issue.

But we could perhaps do slightly better here: since there is only one caller given, we could assign the FILTERS status to the call I think!

File: scout/parse/variant/callers.py [hunk 1]
```diff
@@ -28,36 +36,51 @@ def parse_callers(variant, category="snv"):
...
+                callers[call] = filter_status_default
 
     if raw_info or svdb_origin or other_info:
         return callers
 
     if category == "snv":
-        # cyvcf2 FILTER is None if VCF file column FILTER is "PASS"
         filter_status = "Pass"
         if variant.FILTER is not None:
             filter_status = "Filtered - {}".format(filter_status.replace(";", " - "))
```
Comment (dnil) [inline scout/parse/variant/callers.py:86]: Note that `filter_status` could never take any other value than "Pass" here really, so it was either "Pass" or "Filtered - Pass" I guess?!

File: scout/parse/variant/callers.py [hunk 2]
```diff
@@ -13,6 +13,14 @@ def parse_callers(variant, category="snv"):
     2. If a svdb_origin tag (pipe separated) is found, callers listed will be marked Pass
     3. If a set tag (dash separated, GATK CombineVariants) is found, callers will be marked Pass or Filtered accordingly
 
+    If the FILTER status is not PASS (e.g. None - cyvcf2 FILTER is None if VCF file column FILTER is "PASS"),
```
Comment (dnil) [inline scout/parse/variant/callers.py:16]: I dunno, this function is getting long. If I didn't have the thought about a new parsing module nudging I might split/refactor it. Say if you feel it is way long already.

File: scout/parse/variant/callers.py [hunk 3]
```diff
@@ -28,36 +36,51 @@ def parse_callers(variant, category="snv"):
     svdb_origin = variant.INFO.get("svdb_origin")
     raw_info = variant.INFO.get("set")
 
+    filter_status_default = "Pass"
+    if variant.FILTER is not None:
+        filter_status_default = "Filtered - {}".format(variant.FILTER.replace(";", " - "))
```
Comment (dnil) [inline scout/parse/variant/callers.py:41]: This is one of the crucial changes. As noted below at its origin, the previous version would never add the actual FILTER tag on parsing. Python lazy initiation error. That would just have been a compile error in another language.

File: scout/parse/variant/callers.py [hunk 4]
```diff
@@ -28,36 +36,51 @@ def parse_callers(variant, category="snv"):
...
+
     if other_info:
-        for info in other_info.split(","):
+        infos = other_info.split(",")
+        if len(infos) > 1:
```
Comment (dnil) [inline scout/parse/variant/callers.py:45]: As per the doctoring: add filter status if we are looking at a single caller, else just set as before.

File: scout/server/blueprints/variants/templates/variants/utils.html [hunk 5]
```diff
@@ -56,7 +56,7 @@
       </span>
     {% endfor %}
   {% for name, caller in variant.callers %} <!-- Collect info for specific callers -->
-      <span class="badge {% if caller == 'Pass' %}bg-success{% elif caller == 'Filtered' %}bg-secondary{% else %}bg-black{% endif %}" data-bs-toggle="tooltip" data-bs-html="true" title="{{caller}}">
+      <span class="badge {% if caller == 'Pass' %}bg-success{% elif 'Filtered' in caller %}bg-secondary{% else %}bg-black{% endif %}" data-bs-toggle="tooltip" data-bs-html="true" title="{{caller}}">
```
Comment (dnil) [inline scout/server/blueprints/variants/templates/variants/utils.html:59]: Show if substring matches, as in "Filtered - MinSomaticScore".

**Comment by northwestwitch** [comment]
Just to understand, when there is filter passed but not a specific caller, then originally the filter was not "Passed" but "filtered"?

**Review by northwestwitch — CHANGES_REQUESTED** [review]
This seems to capture more filtered events, as oppsite to what happens on main branch, which is nice! 
Would be nice to: 
- Split the parsing function into different smaller functions, as you also suggested.
- Fix the filter label displayed on variantS page, because it's too long for that cell.

I'm still trying to understand why we still have the green PASS label with or without callers for some cancer variants. I guess I have to look inside the VCF

**Comment by dnil** [comment]
> I'm still trying to understand why we still have the green PASS label with or without callers for some cancer variants. I guess I have to look inside the VCF

Yes, there are some lines there that appear odd to my understanding, much like if it was an amalgamation of different pipeline runs or call sets. I guess the demo is a) old and b) perhaps not entirely representative of one pipeline version, but has some edited lines etc.

**Comment by dnil** [comment]
> Just to understand, when there is filter passed but not a specific caller, then originally the filter was not "Passed" but "filtered"?

Hm, no, I thought it would have been Pass and None on all callers for the category?

File: CHANGELOG.md [hunk 6]
```diff
@@ -10,6 +10,8 @@ About changelog [here](https://keepachangelog.com/en/1.0.0/)
 - Parse ClinVar low-penetrance info and display it alongside Pathogenic and likely pathogenic on SNVs pages
 - Gene panel indexes to reflect the indexes used in production database
 - Panel version check while editing the genes of a panel
+- Display unknown filter tags as "danger" marked badges
+
```
Comment (northwestwitch) [inline CHANGELOG.md:14]:
```suggestion
```

File: scout/parse/variant/callers.py [hunk 7]
```diff
@@ -1,66 +1,140 @@
...
+
+import cyvcf2
 
 from scout.constants import CALLERS
 
 LOG = logging.getLogger(__name__)
 
 
-def parse_callers(variant, category="snv"):
+def parse_callers(variant: cyvcf2.Variant, category: str = "snv") -> dict:
```
Comment (northwestwitch) [inline scout/parse/variant/callers.py:11]: 🏅

File: scout/parse/variant/callers.py [hunk 8]
```diff
@@ -1,66 +1,140 @@
...
-    raw_info = variant.INFO.get("set")
+    def get_callers_from_found_in(info_found_in: str, filter_status: Optional[str]) -> dict:
+        """If a FOUND_IN tag (comma separated) is found, callers listed will be marked Pass.
+        In case of a FILTER status, also set the caller status for a single caller to 'Filtered - status'.
+        If more than one caller, and FILTER status, we cant really tell which said what, and all will be
+        "Filtered".
+        """
+        found_ins: list = info_found_in.split(",")
+
+        filter_status_default = "Pass"
```
Comment (northwestwitch) [inline scout/parse/variant/callers.py:44]: This could also be a constant, outside this function.

File: tests/parse/test_parse_callers.py [hunk 9]
```diff
@@ -46,6 +46,21 @@ def test_parse_callers_complex(cyvcf2_variant):
     assert callers["samtools"] == "Filtered"
 
 
+def test_parse_callers_complex_filter(cyvcf2_variant):
```
Comment (northwestwitch) [inline tests/parse/test_parse_callers.py:49]: 👍🏻

File: tests/parse/test_parse_callers.py [hunk 10]
```diff
@@ -75,6 +90,34 @@ def test_parse_callers_intersection_svdb_info(cyvcf2_variant):
     assert callers["samtools"] is None
 
 
+def test_parse_callers_filtered_svdb_info(cyvcf2_variant):
```
Comment (northwestwitch) [inline tests/parse/test_parse_callers.py:93]: 👍🏻

**Review by northwestwitch — APPROVED** [review]
Nice with the refactoring and new tests! 💯

**Pull request merged**